maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java

```diff
@@ -104,16 +113,53 @@ private TransformerContext getTransformerContext(Map<String, ?> options) {
 
     private Model read(InputStream input, Path pomFile, Map<String, ?> options) throws IOException {
         try {
+            InputSource source = getSource(options);
+            boolean strict = isStrict(options);
             XMLInputFactory factory = new com.ctc.wstx.stax.WstxInputFactory();
-            factory.setProperty(XMLInputFactory.IS_REPLACING_ENTITY_REFERENCES, false);
-            XMLStreamReader parser = factory.createXMLStreamReader(input);
+            XMLOutputFactory outputFactory = new com.ctc.wstx.stax.WstxOutputFactory();
+
+            XMLStreamReader parser;
+            // We only support xml entities and xinclude when reading a file in strict mode
+            if (pomFile != null && strict) {
+                XMLResolver resolver = (String publicID, String systemID, String baseURI, String namespace) -> {
+                    if (systemID == null) {
+                        throw new XMLStreamException("systemID is null");
+                    }
+                    if (baseURI == null) {
+                        throw new XMLStreamException("baseURI is null");
+                    }
+                    URI sysUri;
+                    try {
+                        sysUri = new URI(systemID);
+                    } catch (URISyntaxException e) {
+                        throw new XMLStreamException("Invalid syntax for systemID uri: " + systemID, e);
+                    }
+                    if (sysUri.isAbsolute()) {
+                        throw new XMLStreamException("systemID must be a relative URL: " + systemID);
+                    }
+                    try {
+                        URL url = URI.create(baseURI).resolve(sysUri).toURL();
+                        return new StreamSource(url.openStream(), url.toExternalForm());
+                    } catch (IOException e) {
+                        throw new XMLStreamException("Unable to create Source for " + systemID, e);
+                    }
+                };
+                factory.setProperty(XMLInputFactory.IS_REPLACING_ENTITY_REFERENCES, true);
+                factory.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, true);
+                factory.setProperty(WstxInputProperties.P_ENTITY_RESOLVER, resolver);
+                parser = new XIncludeStreamReader(
+                        factory,
+                        outputFactory,
+                        pomFile.toUri().toString(),
+                        factory.createXMLStreamReader(new StaxPathInputSource(pomFile, input)));
+            } else {
+                parser = factory.createXMLStreamReader(new StreamSource(input));
```

Comment:
Should there be a `factory.setProperty(XMLInputFactory.IS_REPLACING_ENTITY_REFERENCES, false);` call here?

Reply:
The factory is new and `false` is always the default, it should not be necessary.